### PR TITLE
Add extended clipboard Pseudo-Encoding

### LINKED ISSUE #61: Add support for UltraVNCs extended Clipboard
Last year the UltraVNC folks implemented extended clipboard support which provides the ability to handle UTF8 text, RTF and some more formats. It would be nice to see support for this in noVNC too.

### PR DESCRIPTION
**Pull request by CendioNiko**
The current clipboard did not support transferring  Unicode characters. This pull request adds support for extended clipboard pseudo-encoding on the client. The server must also support extended clipboard pseoudo-encoding for this solution to take effect. The servers I know of use this protocol is TigerVNC and UltraVNC.

Fixes #61

### CODE REVIEW

File: tests/test.rfb.js [hunk 1]
```diff
@@ -2335,17 +2348,258 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+
+                    client._sock._websocket._receive_data(new Uint8Array(data));
+
+                    expect(RFB.messages.extendedClipboardProvide).to.have.been.calledOnce;
+                    expect(RFB.messages.extendedClipboardProvide).to.have.been.calledWith(client._sock, expectedData, ["HejHej"]);
+                });
+            });
+        });
+
+        describe('Extended Clipboard Handling Send', function () {
```
Comment (CendioOssman) [inline tests/test.rfb.js]: These are the first tests we have that test `RFB.messages`. I think it might be prudent to create a new grouping that makes this clear.

File: tests/test.deflator.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        let arrayChars = [];
+        for (let i = 0; i < preText.length; i++) {
+            arrayChars[i] = preText.charCodeAt(i);
+        }
+
+        preText = new Uint8Array(arrayChars);
+
+        let compText = deflator.deflate(preText);
+
+        inflator.setInput(compText);
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply tests/test.deflator.js]: As a result of us using pakos inflate() instead of Inflator in our tests the ordering of the commits does not matter anymore.

File: core/rfb.js [hunk 3]
```diff
@@ -1321,22 +1355,175 @@ export default class RFB extends EventTargetMixin {
...
+                    this._clipboardServerCapabilitiesActions[index] = ((actions & index) == index);
+                }
+
+                /*  Caps handling done, send caps with our capabilities
+                    set as a response */
+                RFB.messages.extendedClipboardCaps(this._sock, {extendedClipboardFormatText: 0});
+
+            } else if (actions == extendedClipboardActionRequest) {
+                Log.Debug("Request message received");
+                // Check if server has told us it can handle Provide and there is clipboard data to send.
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply core/rfb.js]: This check is done earlier in this function.

File: core/rfb.js [hunk 4]
```diff
@@ -1321,22 +1355,175 @@ export default class RFB extends EventTargetMixin {
         return this._fail("Unexpected SetColorMapEntries message");
     }
 
+    _count_set_bits_in_range(num, startBit, endBit) {
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply core/rfb.js]: Restructured where it was used and not needed anymore.

File: core/rfb.js [hunk 5]
```diff
@@ -1321,22 +1355,175 @@ export default class RFB extends EventTargetMixin {
...
+        /*  calculating a number 'mask' having
+            endBit number of bits and bits in the
+            range startBit to endBit are the  only set bits */
+        let mask = ((1 << endBit + 1) - 1) ^ ((1 << (startBit)) - 1);
+
+        //  Returns the number of set bits in range startBit to endBit in num.
+        return this._count_set_bits(num & mask);
+    }
+
+    _count_set_bits(num) {
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply core/rfb.js]: Restructured where it was used and not needed anymore.

File: tests/test.rfb.js [hunk 6]
```diff
@@ -2335,17 +2348,258 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+                    client._sock._websocket._receive_data(new Uint8Array(data));
+                });
+
+                it('should update clipboard with correct data from a Provide message', function () {
+                    let expectedData = "Aå漢字!";
+                    let data = [3, 0, 0, 0];
+                    const flags = [0x10, 0x00, 0x00, 0x01];
+
+                    /* The size 10 (utf8 encoded string size) and the
+                    string "Aå漢字!" utf8 encoded and deflated. */
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply tests/test.rfb.js]: No, but we have added another test that tests terminating null. This one can test that we also can handle non null terminated strings.

File: core/rfb.js [hunk 7]
```diff
@@ -1321,22 +1355,175 @@ export default class RFB extends EventTargetMixin {
...
+                    this._clipboardServerCapabilitiesActions[extendedClipboardActionRequest]) {
+
+                    if ((formats & extendedClipboardFormatText) == extendedClipboardFormatText) {
+                        RFB.messages.extendedClipboardRequest(this._sock, [extendedClipboardFormatText]);
+                    }
+                }
+
+            } else if (actions == extendedClipboardActionProvide) {
+                Log.Debug("Provide message received");
+
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply core/rfb.js]: TigerVNC has a FIXME to do this

File: core/rfb.js [hunk 8]
```diff
@@ -1321,22 +1355,175 @@ export default class RFB extends EventTargetMixin {
...
+                if ((formats & extendedClipboardFormatText) != extendedClipboardFormatText) {
+                    return true;
+                }
+                // Ignore what we had in our cliboard client side.
+                this._clipboardText = null;
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply core/rfb.js]: No empty notify is sent when updating from a provide message.

File: core/rfb.js [hunk 9]
```diff
@@ -390,7 +413,17 @@ export default class RFB extends EventTargetMixin {
 
     clipboardPasteFrom(text) {
```
[the remark this answers is not in the document]
Comment (CendioNiko) [reply core/rfb.js]: The API documentation only mentions this under clipboard: "The clipboard event is fired when the server has sent clipboard data. The detail property is an Object containing the property text which is a DOMString with the clipboard data."  Does not seem to be in need of a change.

**Comment by CendioNiko** [comment]
Tested against Xvnc server with extended clipboard pseudo-encoding support on following:
|                              | Windows 10  | Fedora 30 | macOs 10.15.3 |
| ---------------------- | ---------------- | -------------- |-------------------- |
| Chrome                |  ✔ v.79                 | ✔ v.78 | ✔ v.80              |
| Firefox                      | ✔ v.72           | ✔ v.69     | ✔ v.71              |
| Internet Explorer    | ✔ v.11                 | -              | -                        |
| Microsoft Edge  | ✔ v.44                  | -              | -                        |
| Safari                   | -                    | -            | ✔ v.13                      |

Also tested against older Xvnc server that did not support extended clipboard pseudo-encoding on Fedora 30 in Firefox 69.

File: tests/test.rfb.js [hunk 10]
```diff
@@ -292,11 +292,10 @@ describe('Remote Frame Buffer Protocol Client', function () {
 
         describe('#clipboardPasteFrom', function () {
             it('should send the given text in a paste event', function () {
-                const expected = {_sQ: new Uint8Array(11), _sQlen: 0,
-                                  _sQbufferSize: 11, flush: () => {}};
-                RFB.messages.clientCutText(expected, 'abc');
+                sinon.spy(RFB.messages, 'clientCutText');
```
[the remark this answers is not in the document]
Comment (CendioOssman) [reply tests/test.rfb.js]: This issue still exists AFAICT.

File: core/deflator.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    deflate(inData) {
+        this.strm.input = inData;
+        this.strm.avail_in = this.strm.input.length;
+        this.strm.next_in = 0;
+        this.strm.next_out = 0;
+        this.strm.avail_out = this.chunkSize;
+
+        let stream = this.strm;
+
+        // Use Z_FULL_FLUSH be more generic and server bug compatible
```
Comment (CendioOssman) [inline core/deflator.js]: This is a bit too vague. What bug? How do I test for it?

File: core/deflator.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        let lastRet = deflate(stream, Z_FULL_FLUSH);
+        let outData = new Uint8Array(stream.output.buffer, 0, stream.next_out);
+
+        /* If there is more data to read after our chunk we need to
+           continue reading until finished */
```
Comment (CendioOssman) [inline core/deflator.js]: This comment seems misplaced.

File: core/deflator.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            throw new Error("zlib deflate failed");
+        }
+
+        if (stream.avail_in > 0) {
+            // Read chunks until done
+
+            let chunks = [outData];
+            let totalLen = outData.length;
+            do {
+                stream.output = new Uint8Array(this.chunkSize);
```
Comment (CendioOssman) [inline core/deflator.js]: It seems I cannot comment on things GitHub considers "outdated", but this comment doesn't seem to be addressed:

> You need to reset output on the next call to deflate().

File: vendor/pako/lib/zlib/deflate.js [hunk 14]
```diff
@@ -12,7 +12,7 @@ import msg from "./messages.js";
 var Z_NO_FLUSH      = 0;
 var Z_PARTIAL_FLUSH = 1;
 //var Z_SYNC_FLUSH    = 2;
-var Z_FULL_FLUSH    = 3;
+export const Z_FULL_FLUSH    = 3;
```
Comment (CendioOssman) [inline vendor/pako/lib/zlib/deflate.js]: Fix them all if you're at it. :)
(and `inflate.js` as well)

File: core/rfb.js [hunk 15]
```diff
@@ -1325,18 +1364,144 @@ export default class RFB extends EventTargetMixin {
...
+            this.dispatchEvent(new CustomEvent(
+                "clipboard",
+                { detail: { text: text } }));
+
+        } else {
+            //Extended msg.
+            length = Math.abs(length);
+            const flags = this._sock.rQshift32();
+            let formats = flags & 0x0000FFFF;     //0-15
+            let actions = flags & 0xFF000000; //24-31
```
Comment (CendioOssman) [inline core/rfb.js]: Some weird spacing of the comments here. And perhaps drop the comments? People should be able to deduce the bits easily from those masks.

File: core/rfb.js [hunk 16]
```diff
@@ -1325,18 +1364,144 @@ export default class RFB extends EventTargetMixin {
...
         this._sock.rQskipBytes(3);  // Padding
-        const length = this._sock.rQshift32();
-        if (this._sock.rQwait("ServerCutText", length, 8)) { return false; }
 
-        const text = this._sock.rQshiftStr(length);
+        let length = this._sock.rQshift32();
+        length = toSigned32bit(length);
+
+        if (this._sock.rQwait("ServerCutText content", Math.abs(length), 8)) { return false; }
+        if (this._viewOnly) {
```
Comment (CendioOssman) [inline core/rfb.js]: I'm afraid this is incorrect. View only can be toggled while connected, so we need to handle messages such as "caps" in case the user disables view only.

File: core/rfb.js [hunk 17]
```diff
@@ -390,7 +413,22 @@ export default class RFB extends EventTargetMixin {
 
     clipboardPasteFrom(text) {
         if (this._rfb_connection_state !== 'connected' || this._viewOnly) { return; }
-        RFB.messages.clientCutText(this._sock, text);
+
+        if (this._clipboardServerCapabilitiesFormats != null &&
```
Comment (CendioOssman) [inline core/rfb.js]: I thought we got rid of these null checks?

File: core/rfb.js [hunk 18]
```diff
@@ -1325,18 +1364,144 @@ export default class RFB extends EventTargetMixin {
...
+                        let sizeArray = streamInflator.inflate(4);
+
+                        size |= (sizeArray[0] << 24);
+                        size |= (sizeArray[1] << 16);
+                        size |= (sizeArray[2] << 8);
+                        size |= (sizeArray[3]);
+                        chunk = streamInflator.inflate(size);
+                    }
 
+                    if ((((1 << i) & formats) == extendedClipboardFormatText) && (chunk != null)) {
```
Comment (CendioOssman) [inline core/rfb.js]: I think these can be combined even more. And `chunk` is never `null`, so there is a bug here.

File: core/rfb.js [hunk 19]
```diff
@@ -1325,18 +1364,144 @@ export default class RFB extends EventTargetMixin {
...
+                        textData = String.fromCharCode.apply(null, chunk);
+
+                        textData = decodeUTF8(textData);
+                        if ((textData.length > 0) && "\0" == textData.charAt(textData.length - 1)) {
```
Comment (CendioOssman) [inline core/rfb.js]: Try to use `===` in JavaScript to avoid bugs.

File: core/rfb.js [hunk 20]
```diff
@@ -1325,18 +1364,144 @@ export default class RFB extends EventTargetMixin {
...
+                            textData = textData.slice(0, -1);
+                        }
+
+                        textData =textData.replace("\r\n", "\n");
```
Comment (CendioOssman) [inline core/rfb.js]: There's a missing space here.

File: core/rfb.js [hunk 21]
```diff
@@ -1966,8 +2131,102 @@ RFB.messages = {
...
+        let dataToDeflate = [];
+
+        for (let i = 0; i < formats.length; i++) {
+            // We only support the format Text at this time
+            if (formats[i] != extendedClipboardFormatText) {
+                continue;
+            }
+
+            // Change lone \r or \n into \r\n as defined in rfbproto
+            inData[i] = inData[i].replace(/\r\n|\r|\n/gm,   "\r\n");
```
Comment (CendioOssman) [inline core/rfb.js]: Some stray whitespace here.

File: core/rfb.js [hunk 22]
```diff
@@ -1966,8 +2131,102 @@ RFB.messages = {
...
+
+    extendedClipboardProvide(sock, formats, inData) {
+        // Deflate incomming data and their sizes
+        let deflator = new Deflator();
+        let dataToDeflate = [];
+
+        for (let i = 0; i < formats.length; i++) {
+            // We only support the format Text at this time
+            if (formats[i] != extendedClipboardFormatText) {
+                continue;
```
Comment (CendioOssman) [inline core/rfb.js]: But we still send this `formats` to the server. So we need to be harsher here. I'd throw an exception.

File: tests/test.rfb.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (CendioNiko) [reply tests/test.rfb.js]: Forgot to add this change in earlier commit. Fixed now.

File: tests/test.rfb.js [hunk 23]
```diff
@@ -2573,3 +2832,166 @@ describe('Remote Frame Buffer Protocol Client', function () {
...
+            expectedData[3] = 0x01; // The formats client supports
+            expectedData.set(deflatedData, 4);
+
+            RFB.messages.extendedClipboardProvide(sock, [0x01], [testText]);
+            expect(RFB.messages.clientCutText).to.have.been.calledOnce;
+            expect(RFB.messages.clientCutText).to.have.been.calledWith(sock, expectedData, true);
+
+        });
+
+        describe('End of line characters', function () {
```
Comment (CendioOssman) [inline tests/test.rfb.js]: Really nice coverage here. Well done!

File: core/rfb.js (hunk 20 above)
Comment (samhed) [reply core/rfb.js]: this seems like something eslint should have cought

File: core/deflator.js [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+                offset += chunks[i].length;
+            }
+
+            outData = newData;
+        }
+
+        this.strm.input = null;
+        this.strm.avail_in = 0;
+        this.strm.next_in = 0;
+
```
Comment (samhed) [inline core/deflator.js]: remove one newline here

File: core/rfb.js [hunk 25]
```diff
@@ -6,12 +6,14 @@
  * See README.md for usage and integration instructions.
  *
  */
-
```
Comment (samhed) [inline core/rfb.js]: This should not have been removed

Comment (samhed) [reply core/rfb.js]: And with this big diff to rfb.js, I'd update the 'Copyright'-year as well

File: core/rfb.js [hunk 26]
```diff
@@ -84,6 +103,10 @@ export default class RFB extends EventTargetMixin {
 
         this._qemuExtKeyEventSupported = false;
 
+        this._clipboardServerCapabilitiesFormats = null;
```
Comment (samhed) [inline core/rfb.js]: If you initiate this containing all the formats as false from the start, you will not have to do all the null checks @CendioOssman talks about below
